odl/discr/discr_mappings.py

```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
                                                   values.ndim))
-
-        # Cast to floating point was removed here
-
-        # TODO: check if the following code is needed
-
-#        self.fill_value = fill_value
-#        if fill_value is not None:
-#            fill_value_dtype = np.asarray(fill_value).dtype
-#            if (hasattr(values, 'dtype')
-#                    and not np.can_cast(fill_value_dtype, values.dtype,
-#                                        casting='same_kind')):
-#                raise ValueError("fill_value must be either 'None' or "
-#                                 "of a type compatible with values")
-
         for i, p in enumerate(coord_vecs):
             if not np.asarray(p).ndim == 1:
                 raise ValueError('The points in dimension {} must be '
                                  '1-dimensional'.format(i))
-            if not values.shape[i] == len(p):
+            if values.shape[i] != len(p):
                 raise ValueError('There are {} points and {} values in '
                                  'dimension {}'.format(len(p),
                                                        values.shape[i], i))
 
-        # TODO: use the following code to determine upstream if a different
-        # interpolator needs to be used
-
-#        for i, p in enumerate(points):
-#            if not np.all(np.diff(p) > 0.):
-#                raise ValueError("The points in dimension {} must be strictly"
-#                                 " ascending".format(i))
         self.grid = tuple([np.asarray(p) for p in coord_vecs])
         self.values = values
+        self.type = typ
 
     def __call__(self, x, out=None):
         """Do the interpolation.
 
-        Modified for in-place evaluation support and without method
-        choice. Evaluation points are to be given as an array with
-        shape (n, dim), where n is the number of points.
+        Parameters
+        ----------
+        x : meshgrid or `numpy.ndarray`
+            Evaluation points of the interpolator
+        out : `numpy.ndarray`, optional
+            Array to which the results are written. Needs to have
+            correct shape according to input ``x``.
+
+        Returns
+        -------
+        out : `numpy.ndarray`
+            Interpolated values. If ``out`` was given, the returned
+            object is a reference to it.
         """
-        ndim = len(self.grid)
-        if x.ndim != 2:
-            raise ValueError('x has {} axes instead of 2.'.format(x.ndim))
-
-        if x.shape[0] != ndim:
-            raise ValueError('x has axis 1 with length {} instead '
-                             'of the grid dimension {}.'
-                             ''.format(x.shape[0], ndim))
+        if self.type == 'array':
+            # Make a (1, n) array from one with shape (n,)
+            ndim = len(self.grid)
+            if ndim == 1 and x.ndim == 1:
+                x = x[None, :]
```

Comment:
I would use `x = x.reshape([ndim, -1])` as it communicates the goal better. Then you could also remove the two following tests since this would fail and we can rely on numpys testing instead of writing our own.